Book transaction

esterinac opened the pull request:
### Book Transaction
- Mencatat log transaksi buku masuk dan keluar gudang
- Buku masuk gudang:
   - Percetakan (dicatat di transaksi buku masuk saat finalisasi penerimaan buku)
- Buku keluar gudang: 
   - Pesanan (faktur)
   - Pemindahan (dari gudang ke perpustakaan/showroom) (belum di PR)
   - Non penjualan (doorprize/present gift) (belum di PR)
- Saat memilih judul buku di halaman index, pengguna akan diarahkan ke view detail penerimaan/pesanan/pemindahan/non penjualan dari buku tersebut.
![screencapture-localhost-sigap-book-transaction-2021-05-15-10_39_41](https://user-images.githubusercontent.com/66486312/118346983-e5692980-b569-11eb-8955-473d462d7b39.png)
### Migration
- Create table book_transaction untuk catat log transaksi buku masuk/keluar gudang
- Contoh data:
![image](https://user-images.githubusercontent.com/66486312/118346974-d08c9600-b569-11eb-8bcc-98f292c8b4ac.png)
- Kalau buku masuk gudang, insert data stock_in dan book_receive_id
- Kalau buku keluar, insert data stock_out dan invoice_id/book_transfer_id/book_non_sales_id

esterinac committed:
Book Transaction

esterinac committed:
Insert book receive to book transaction

esterinac committed:
Fix transaction date on book receive

esterinac committed:
Fix index layout

bagaskarala commented on application/modules/book_transaction/controllers/Book_transaction.php:
```diff
@@ -0,0 +1,163 @@
+    }
+
+
+    public function view($book_transaction_id){
+        // $book_transaction = $this->book_transaction->join('book')->where('book.book_id', $book_id)->get();
+        $book_transaction = $this->book_transaction->get_book_transaction($book_transaction_id);
+        if (!$book_transaction) {
+            $this->session->set_flashdata('warning', $this->lang->line('toast_data_not_available'));
+            redirect($this->pages);
+        }
+
+        $input = (object) $book_transaction;
+        // $get_transaction      = $this->book_transaction->fetch_transaction_by_id($book_transaction_id);
+        // $transaction_history  = $get_transaction['transaction_history'];
+        // $transaction_last     = $get_transaction['transaction_last'];
+
+        $pages       = $this->pages;
+        $main_view   = 'book_transaction/view_booktransaction';
+        $this->load->view('template', compact('pages', 'main_view', 'input'));
+        return;
+    }
```
delete unused code in comment

esterinac replied:
oke mas

bagaskarala commented on application/modules/book_transaction/models/Book_transaction_model.php:
```diff
@@ -0,0 +1,96 @@
+    public $per_page = 10;
+    public function filter_book_transaction($filters, $page)
+    {
+        $book_transactions = $this->select([ 
+            'print_order.print_order_id','print_order.order_number',
+            'book_receive.book_receive_id',
+            'invoice.invoice_id', 'invoice.number as invoice_number', 
+            'book_non_sales.book_non_sales_id', 'book_non_sales.number as book_non_sales_number',
+            'book_transfer.book_transfer_id', 'book_transfer.transfer_number',         
+            'book.book_title', 'book_transaction.*'])
+            ->where_not('date', NULL)
+            ->join_table('book', 'book_transaction', 'book')
+            ->join_table('book_receive', 'book_transaction', 'book_receive')
+            ->join_table('print_order', 'book_receive', 'print_order')
+            ->join_table('invoice', 'book_transaction', 'invoice')
+            ->join_table('book_transfer', 'book_transaction', 'book_transfer')
+            ->join_table('book_non_sales', 'book_transaction', 'book_non_sales')
+            ->when('keyword', $filters['keyword'])
+            ->when('start_date', $filters['start_date'])
+            ->when('end_date', $filters['end_date'])
+            ->when('transaction_type', $filters['transaction_type'])
+            ->order_by('book_transaction_id', 'DESC')
+            ->paginate($page)
+            ->get_all();
```
yang tabelnya belum ada, didisable dulu join.nya karena bikin error.

![image](https://user-images.githubusercontent.com/14112952/117906445-1bef3c00-b2ff-11eb-9fda-9ef89eb0f799.png)

bagaskarala replied:
atau jika PR ini tergantung dgn fitur yg lain, bikin PR yang lain tsb dulu aja

esterinac replied:
oh iya, aku comment dulu aja ya mas select sama join tabel-tabel yang belum ada sekarang. nanti pas bikin PR fitur yang lain, yang di sini aku uncomment lagi. soalnya harus ada fitur ini buat nyatet buku masuk/keluar gudang

bagaskarala replied:
ini book transaction cara alur demonya gimana?

esterinac replied:
kalau yang saat ini:
buat order cetak baru->penerimaan buku->pas finalisasi nanti insert ke tabel book transaction buat catet buku masuk gudang mas

esterinac committed:
Delete unused code

esterinac committed:
Disable select and join table book_transfer & book_non_sales (blm ada tabelnya)

bagaskarala commented:
oiya mau nanya, ini kalo pas bikin faktur tipe tunai-perpustakaan di pesanan buku tidak ada progressnya? @esterinac @michaelr1300 
![image](https://user-images.githubusercontent.com/14112952/117934015-633ff180-b32c-11eb-9151-99bdf811dbf2.png)

bagaskarala commented on application/modules/book_receive/controllers/Book_receive.php:
```diff
@@ -465,9 +465,22 @@ public function final($book_receive_id = null, $action = null)
         ]);
 
-        // update book stock
         $book_stock = $this->book_stock->where('book_id', $book_receive->book_id)->get();
         $book_stock_print = $this->book_receive->get_print_order($book_receive->print_order_id);
+
+        //insert to book transaction
+        $book_stock = $this->book_stock->where('book_id', $book_receive->book_id)->get();
```
double `$book_stock`

ketika klik finalisasi penerimaan buku, maka muncul error seperti gambar di bawah.
oiya kenapa perlu `/finish` di belakang? bukannya kalo finalisasi book receive udah pasti statusnya di set ke `finish`? atau ada kondisi lain?

sepertinya ini error karena `book_stock`nya ga ada, nah `book_stock` itu ditambahkan saat kapan?

![image](https://user-images.githubusercontent.com/14112952/117935470-ef9ee400-b32d-11eb-8da6-c6285d961937.png)

esterinac replied:
oh iya mas nanti dihapus `$book_stock` yang double.
harusnya tidak perlu /finish sih mas karena emang status akhirnya selalu finish, aku ganti ya mas biar langsung /final/id aja.
di tempatku bisa mas, mungkin karena di database mas bagas ada beberapa baris yang book_id nya dobel/sama? karena harusnya book sama book_stock relasinya one to one. mungkin bisa dihapus dulu mas data yang book_id nya ke-double

bagaskarala replied:
@esterinac nah data book_stock itu ditambahkan/insert ketika apa?

esterinac replied:
saat final mas

bagaskarala replied:
coba cek method `final` di controller `book_receive`, kayaknya urutannya kebalik. insert `book_transaction` harusnya setelah pengecekan `book_stock` kan?

cara reproduce errornya gini, coba order cetak buku yang belum ada di book stock.

bagaskarala replied:
harusnya yg bener kayak gini bukan?

https://github.com/bagaskarala/sigap/blob/develop/application/modules/book_receive/controllers/Book_receive.php#L469-L488

esterinac replied:
oh iya mas kalau cetak buku yang belum ada di book stock jadinya error, karena belum ada `book_stock_id` nya ya. iya mas kebalik ternyata bener yang sebelumnya, makasih mas. sudah kubenerin lagi

esterinac commented:
> oiya mau nanya, ini kalo pas bikin faktur tipe tunai-perpustakaan di pesanan buku tidak ada progressnya? @esterinac @michaelr1300
> ![image](https://user-images.githubusercontent.com/14112952/117934015-633ff180-b32c-11eb-9151-99bdf811dbf2.png)

tidak ada progressnya mas, yang ada progressnya cuma faktur yang ambil bukunya dari gudang

esterinac commented:
> oiya mau nanya, ini kalo pas bikin faktur tipe tunai-perpustakaan di pesanan buku tidak ada progressnya? @esterinac @michaelr1300
> ![image](https://user-images.githubusercontent.com/14112952/117934015-633ff180-b32c-11eb-9151-99bdf811dbf2.png)

tidak ada progressnya mas, yang ada progressnya cuma faktur yang ambil bukunya dari gudang. 
oh iya mas harusnya yang muncul di pesanan itu cuma faktur yang asal pengambilan buku (source) nya dari gudang, karena pesanan buku fungsinya untuk menyiapkan buku keluar dari gudang karena faktur. nanti aku benerin lagi ya mas

esterinac committed:
Fix pesanan buku

-Menampilkan faktur yang memiliki asal/source warehouse
- Tipe faktur di view

esterinac committed:
Fix penerimaan buku

- fix double $book_stock
- fix final (final/id)

esterinac committed:
Fix book receive final

bagaskarala commented:
cara sinkronisasi stok_book_revision dengan book_transaction gmn? karena bisa beda jika stok buku diubah, sedangkan di transaksi tidak ada

esterinac commented:
> cara sinkronisasi stok_book_revision dengan book_transaction gmn? karena bisa beda jika stok buku diubah, sedangkan di transaksi tidak ada

book_stock_revision dengan book_transaction itu beda mas, kalau book_stock_revision itu pengubahan jumlah stok karena ada perbedaan antara jumlah stok yang tercatat dengan jumlah stok yang sebenarnya ada di gudang, biasanya baru terlihat saat dilakukan pengecekan stok. kalau book_transaction itu untuk mencatat buku yang resmi masuk/keluar gudang seperti cetak, jual, doorprize.

mas bagas nanya ini karena ada stok awal sama stok akhir di halaman indexnya kah? kalau karena itu, lebih baik kalo di transaksi buku hanya menampilkan perubahan stoknya saja ya mas seperti ini? 
![image](https://user-images.githubusercontent.com/66486312/118072035-398dd580-b3d3-11eb-83b4-c3b776c8a7f1.png)
soalnya kemarin juga sudah sempet demo di ugm press dan di transaksi buku hanya menampilkan perubahan stoknya (tidak ada stok awal dan akhir) seperti ini, dan kemarin juga sudah pada oke kalau transaksi bukunya begini.

bagaskarala commented:
> 
> book_stock_revision dengan book_transaction itu beda mas, kalau book_stock_revision itu pengubahan jumlah stok karena ada perbedaan antara jumlah stok yang tercatat dengan jumlah stok yang sebenarnya ada di gudang, biasanya baru terlihat saat dilakukan pengecekan stok. kalau book_transaction itu untuk mencatat buku yang resmi masuk/keluar gudang seperti cetak, jual, doorprize.
> 
> mas bagas nanya ini karena ada stok awal sama stok akhir di halaman indexnya kah? kalau karena itu, lebih baik kalo di transaksi buku hanya menampilkan perubahan stoknya saja ya mas seperti ini?
> ![image](https://user-images.githubusercontent.com/66486312/118072035-398dd580-b3d3-11eb-83b4-c3b776c8a7f1.png)
> soalnya kemarin juga sudah sempet demo di ugm press dan di transaksi buku hanya menampilkan perubahan stoknya (tidak ada stok awal dan akhir) seperti ini, dan kemarin juga sudah pada oke kalau transaksi bukunya begini.

nah kalau book_transaction hanya mencatat perbuahannya berarti tidak perlu mencatat kondisi stoknya juga, karena mutasi stok udah dihandle di book_stock_revision. setiap update book_transaction, juga update book_stock_revision juga kan? berarti book_transaction cukup mencatat jumlah perubahannya saja, tetapi data yang mencatat mutasi stock tetap di book_stock_revision biar konsisten.


tambahan:
atau malah book_transaction tidak perlu mencatat ttg stock sama sekali, karena ada referensi ke book_stock_id kan bisa dijoin (?)

esterinac commented:
> > book_stock_revision dengan book_transaction itu beda mas, kalau book_stock_revision itu pengubahan jumlah stok karena ada perbedaan antara jumlah stok yang tercatat dengan jumlah stok yang sebenarnya ada di gudang, biasanya baru terlihat saat dilakukan pengecekan stok. kalau book_transaction itu untuk mencatat buku yang resmi masuk/keluar gudang seperti cetak, jual, doorprize.
> > mas bagas nanya ini karena ada stok awal sama stok akhir di halaman indexnya kah? kalau karena itu, lebih baik kalo di transaksi buku hanya menampilkan perubahan stoknya saja ya mas seperti ini?
> > ![image](https://user-images.githubusercontent.com/66486312/118072035-398dd580-b3d3-11eb-83b4-c3b776c8a7f1.png)
> > soalnya kemarin juga sudah sempet demo di ugm press dan di transaksi buku hanya menampilkan perubahan stoknya (tidak ada stok awal dan akhir) seperti ini, dan kemarin juga sudah pada oke kalau transaksi bukunya begini.
> 
> nah kalau book_transaction hanya mencatat perbuahannya berarti tidak perlu mencatat kondisi stoknya juga, karena mutasi stok udah dihandle di book_stock_revision. setiap update book_transaction, juga update book_stock_revision juga kan? berarti book_transaction cukup mencatat jumlah perubahannya saja, tetapi data yang mencatat mutasi stock tetap di book_stock_revision biar konsisten.
> 
> tambahan:
> atau malah book_transaction tidak perlu mencatat ttg stock sama sekali, karena ada referensi ke book_stock_id kan bisa dijoin (?)

oh maksud mas bagas tuh berarti harus ada 1 tabel yang bisa mencatat semua mutasi buku (mau revisi atau pun transaksi) gitu kah? kalau yg sekarang sih perubahan data stoknya masih kepisah mas, kalau revisi ya insert ke tabel revisi, kalau transaksi resmi masuk/keluar ya insert ke tabel transaksi. jadi untuk yang sekarang ini, tiap update book_transaction tidak update book_stock_revision mas.

kalau aku tambahin kolom book_stock_revision_id di tabel book_transaction gimana mas? jadi semua mutasi (entah revisi/transaksi) bisa kecatet di situ semua nya, supaya kami juga tidak perlu mengubah banyak.

kalau ada referensi ke book_stock_id jadi bisa dijoin, berarti kolom stock_initial sama stock_last di tabel book_transaction dihapus aja ya. aku ubah di file migratenya ya mas.

bagaskarala commented:
> oh maksud mas bagas tuh berarti harus ada 1 tabel yang bisa mencatat semua mutasi buku (mau revisi atau pun transaksi) gitu kah? kalau yg sekarang sih perubahan data stoknya masih kepisah mas, kalau revisi ya insert ke tabel revisi, kalau transaksi resmi masuk/keluar ya insert ke tabel transaksi. jadi untuk yang sekarang ini, tiap update book_transaction tidak update book_stock_revision mas.
> 
> kalau aku tambahin kolom book_stock_revision_id di tabel book_transaction gimana mas? jadi semua mutasi (entah revisi/transaksi) bisa kecatet di situ semua nya, supaya kami juga tidak perlu mengubah banyak.
> 
> kalau ada referensi ke book_stock_id jadi bisa dijoin, berarti kolom stock_initial sama stock_last di tabel book_transaction dihapus aja ya. aku ubah di file migratenya ya mas.

concernku tuh karena skrg ada pencatatan stock di 2 tempat, itu berpotensi bikin data gak konsisten. harusnya data stock ya cuma ada di satu sumber aja.

hooh join ke book_stock utk dapetin data stock.
lalu join ke book_stock_revision utk dapetin data mutasi.
dicoba dulu ya trus dicek2 beberapa case-nya

esterinac committed:
Fix book transaction

- Add column book_stock_revision_id on table book_transaction
- Get stock changes/mutation from other tables
- join_table_id on MY_MODEL (add parameter id on joining tables)

esterinac commented:
> > oh maksud mas bagas tuh berarti harus ada 1 tabel yang bisa mencatat semua mutasi buku (mau revisi atau pun transaksi) gitu kah? kalau yg sekarang sih perubahan data stoknya masih kepisah mas, kalau revisi ya insert ke tabel revisi, kalau transaksi resmi masuk/keluar ya insert ke tabel transaksi. jadi untuk yang sekarang ini, tiap update book_transaction tidak update book_stock_revision mas.
> > kalau aku tambahin kolom book_stock_revision_id di tabel book_transaction gimana mas? jadi semua mutasi (entah revisi/transaksi) bisa kecatet di situ semua nya, supaya kami juga tidak perlu mengubah banyak.
> > kalau ada referensi ke book_stock_id jadi bisa dijoin, berarti kolom stock_initial sama stock_last di tabel book_transaction dihapus aja ya. aku ubah di file migratenya ya mas.
> 
> concernku tuh karena skrg ada pencatatan stock di 2 tempat, itu berpotensi bikin data gak konsisten. harusnya data stock ya cuma ada di satu sumber aja.
> 
> hooh join ke book_stock utk dapetin data stock.
> lalu join ke book_stock_revision utk dapetin data mutasi.
> dicoba dulu ya trus dicek2 beberapa case-nya

untuk mendapatkan data stok awal/akhir pada tiap transaksi/revisi/tanggal pake join kayaknya kurang memungkinkan mas, susah dapetin data stok gudang pada saat transaksi/revisi tertentu. kalau dapetin data mutasi/perubahannya sudah bisa mas. jadinya di halaman index seperti ini, hanya menampilkan perubahannya. yang tipe pesanan aku masukkin ke transaksi secara manual karena dari fitur faktur memang belum ada insert ke book_transaction.
![image](https://user-images.githubusercontent.com/66486312/118165885-0ab54500-b44f-11eb-8cff-73087b418bfb.png)

esterinac commented on application/core/MY_Model.php:
```diff
@@ -249,6 +249,23 @@ public function join_table($table_to, $table_from, $column, $type = 'left')
+     * @param string $table_dest
+     * @param string $table_middle
+     * @return this
+     **/
```
sama tadi juga nambahin fungsi `join_table_id` buat isi id nya sendiri di MY_MODEL, karena nama PK di tabel invoice_book itu id, bukan invoice_book_id

esterinac commented on application/core/MY_Model.php:
```diff
@@ -249,6 +249,23 @@ public function join_table($table_to, $table_from, $column, $type = 'left')
         return $this;
     }
 
+    /**
+     * Menggabungkan tabel secara fleksibel dan bisa isi nama id
+     *
+     * contoh:
+     * input: join_table_id('invoice_book', 'book_transaction', 'id', 'invoice_book_id')
+     * output: join('invoice_book','invoice_book.id == book_transaction.invoice_book_id')
+     *
+     * @param string $table_dest
+     * @param string $table_middle
+     * @return this
+     **/
+    public function join_table_id($table_to, $table_from, $id_to, $id_from, $type = 'left')
+    {
+        $this->db->join($table_to, "$table_to.{$id_to} = $table_from.{$id_from}", $type);
+        return $this;
+    }
+
```
sama tadi juga nambahin fungsi join_table_id buat isi id nya sendiri di MY_MODEL, karena nama PK di tabel invoice_book itu id, bukan invoice_book_id

bagaskarala replied:
ini `invoice_book_id` (?) https://github.com/bagaskarala/sigap/blob/5ca0b3c64c3c6469bf4dbd18f387f7f2a1be00cb/application/database/migrations/20210501145022_Invoice_book.php#L8

esterinac replied:
oh iya mas aku salah migrate kayanya, sudah kubenerin lagi mas buat ambil invoice_book_id di modelnya

bagaskarala commented:
> untuk mendapatkan data stok awal/akhir pada tiap transaksi/revisi/tanggal pake join kayaknya kurang memungkinkan mas, susah dapetin data stok gudang pada saat transaksi/revisi tertentu. kalau dapetin data mutasi/perubahannya sudah bisa mas. jadinya di halaman index seperti ini, hanya menampilkan perubahannya. yang tipe pesanan aku masukkin ke transaksi secara manual karena dari fitur faktur memang belum ada insert ke book_transaction.

oh ya oke,, kalo udah dijoin ke `book_stock_revision` kan bisa dapetin data perubahan stok dan data stok saat itu juga..
berarti `book_transaction` sudah tidak perlu dijoin ke `book_stock` lagi ya ?

"yang tipe pesanan aku masukkin ke transaksi secara manual" ini maksudnya gimana ya @esterinac ?

bagaskarala commented on application/database/migrations/20210511171916_Book_transaction.php:
```diff
@@ -17,7 +17,7 @@ public function up(){
                 'constraint' => 10,
             ],
-            'invoice_id' => [
+            'invoice_book_id' => [
```
ini kok ke `invoice_book` utk apa ya? gak ke `invoice`?

esterinac replied:
karena kalau di `invoice` itu buku yang keluar tiap faktur bisa lebih dari satu buku mas. misal tiap ada faktur yang dibuat, ada 1 row di tabel `invoice` ditambahkan dan bisa ada beberapa row `invoice_book` ditambahkan. contohnya seperti ini.
tabel invoice:
![image](https://user-images.githubusercontent.com/66486312/118346220-0890da80-b564-11eb-8b99-ff554e7f13a3.png)
tabel invoice_book:
![image](https://user-images.githubusercontent.com/66486312/118346239-25c5a900-b564-11eb-94ae-12c426575454.png)
untuk mendapatkan jumlah buku (qty) di suatu invoice menggunakan join jadinya pake id dari tabel invoice_book (list buku faktur) supaya lebih mudah ambil qty per buku di faktur.

bagaskarala replied:
relate kesini https://github.com/bagaskarala/sigap/pull/42#discussion_r632945606

bagaskarala commented on application/database/migrations/20210511171916_Book_transaction.php:
```diff
@@ -27,39 +27,25 @@ public function up(){
                 'type' => 'INT',
                 'constraint' => 10,
                 'null' => TRUE
             ],
```
ini namenya pake "list" karena apa?

esterinac replied:
sama kayak yang di faktur mas, jadi tiap pemindahan buku dan buku non penjualan (doorprize/present gift) itu bisa keluar lebih dari 1 buku. lebih mudah ambil qty/jumlah perubahannya kalau pake id dari tabel list bukunya

bagaskarala replied:
kalo menurutku mending gini aja, ketika finishing receive/transfer/invoice/non_sales itu langsung sekalian insert mutasi di `book_stock_revision` dan update/insert stock terakhir di `book_stock`. kemudian dilanjutkan insert di `book_transaction`.
nah di `book_transaction` kan ada referensi ke `book_stock_revision` jadi bisa ambil perubahan stoknya utk ditampilkan di index transaksi. sehingga tdk perlu lagi ngambil data qty di invoice, book_receive, dll. karena patokannya dari data perubahan mutasi aja. 

gimana kira2 kalo begitu? ada case yg blm tercover?

esterinac replied:
maksud mas jadinya stok awal sama stok akhir gudang tetep harus ada dan ditampilin kah mas? 

mas kalo misal saat finishing receive/transfer/invoice/non_sales/revisi itu insertnya ke `book_transaction` aja gimana? soalnya kalau `book_stock_revision` kan emang tujuannya khusus untuk revisi aja mas, kayaknya kurang tepat kalau insert receive/transfer/invoice/non_sales ke tabel `book_stock_revision`(?) jadinya semua mutasi diinsert ke `book_transaction` aja.

terus stok awal dan terakhirnya juga insert di `book_transaction` juga kah mas?

bagaskarala replied:
owhh oke, brrti data mutasi yang utama yg `book_transaction` ya, aku tdi mikirnya yg utama `book_stock_revision`. kalo `book_stock_revision` hanya fokus ke revisi stock aja brrti nanti ketika update `book_stock_revision` maka sekaligus update `book_stock` dan insert di `book_transaction` juga kan?
oke coba diimplementasi mutasi di `book_transaction`, ntar ku tes2 lagi

bagaskarala replied:
iya kalo gitu berarti stok awal dan akhir kecatet di `book_transaction`

esterinac replied:
oke mas udah, jadinya gini
![image](https://user-images.githubusercontent.com/66486312/118366019-378d6780-b5c9-11eb-8c13-ad0439c546d6.png)
![screencapture-localhost-sigap-book-transaction-2021-05-15-21_59_00](https://user-images.githubusercontent.com/66486312/118365921-cc439580-b5c8-11eb-82ec-bbbb3e698be8.png)
saat pembuatan faktur/`add()`, stok gudang diupdate dan data transaksi diinsert. stok gudang diupdate saat `add()` supaya keliatan stok yang benar-benar available di gudang buat dijual/dikeluarkan ada berapa. kalau fakturnya ditolak, nanti transaksinya dihapus dan stoknya dibalikkan ke semula lagi. untuk update stok di `edit()` invoice/faktur nanti dikerjain sama @michaelr1300

esterinac commented:
> > untuk mendapatkan data stok awal/akhir pada tiap transaksi/revisi/tanggal pake join kayaknya kurang memungkinkan mas, susah dapetin data stok gudang pada saat transaksi/revisi tertentu. kalau dapetin data mutasi/perubahannya sudah bisa mas. jadinya di halaman index seperti ini, hanya menampilkan perubahannya. yang tipe pesanan aku masukkin ke transaksi secara manual karena dari fitur faktur memang belum ada insert ke book_transaction.
> 
> oh ya oke,, kalo udah dijoin ke `book_stock_revision` kan bisa dapetin data perubahan stok dan data stok saat itu juga..
> berarti `book_transaction` sudah tidak perlu dijoin ke `book_stock` lagi ya ?
> 
> "yang tipe pesanan aku masukkin ke transaksi secara manual" ini maksudnya gimana ya @esterinac ?

tidak perlu mas
yang tipe pesanan aku masukkin lewat phpmyadmin. jadi aku bikin faktur terus aku liat id invoice_book nya, book_id, book_stock_id nya dll terus aku masukkin ke tabel book transaction buat ngecek kalau ada transaksi pesanan/faktur, buku dan jumlah perubahannya bisa ditampilin/tidak di transaksi.

esterinac committed:
Fix invoice_book_id

esterinac committed:
Fix generate excel book transaction

esterinac committed:
Fix generate excel book transaction;

bagaskarala commented:
> tidak perlu mas
> yang tipe pesanan aku masukkin lewat phpmyadmin. jadi aku bikin faktur terus aku liat id invoice_book nya, book_id, book_stock_id nya dll terus aku masukkin ke tabel book transaction buat ngecek kalau ada transaksi pesanan/faktur, buku dan jumlah perubahannya bisa ditampilin/tidak di transaksi.

owalah maksudnya manual cuma buat tes ya. oke deh berarti perlu sekalian dibuat mekanisme inserting-nya, nanti agar sesuai alur yang aku tulis disini https://github.com/bagaskarala/sigap/pull/42#discussion_r632945606

esterinac committed:
Fix book transaction

Fix book_transaction migration file
Insert invoice to transaction

michaelr1300 committed:
Update book transaction saat edit_invoice

- Update view edit_invoice
- Update book transaction saat edit invoice / cancel invoice

michaelr1300 commented on application/modules/invoice/views/edit_invoice.php:
```diff
@@ -23,17 +23,25 @@
+                                        <td><?= get_invoice_type()[$invoice->type]; ?></td>
+                                    </tr>
+                                </tbody>
+                            </table>
```
Jenis faktur tidak dapat diubah (karena akan mempengaruhi nomor faktur)

michaelr1300 commented on application/modules/invoice/controllers/Invoice.php:
```diff
@@ -192,12 +189,39 @@ public function edit($invoice_id)
+                $book_stock = $this->book_stock->where('book_id', $invoice_book->book_id)->get();
+                $book_stock->warehouse_present += $invoice_book->qty;
+                $this->book_stock->where('book_id', $invoice_book->book_id)->update($book_stock);
+            }
```
Stock buku di invoice_book dikembalikan ke book_stock (Karena stock dikurangi saat add_invoice).
Sistem edit:

1. Faktur memiliki invoice_book lama (hasil dari add_invoice, sebelum di edit). 
2. User mengedit invoice_book di faktur, submit invoice_book baru (hasil editan)
3. Semua invoice_book lama di hapus, stock dikembalikan ke data book_stock
4. invoice_book baru ditambahkan sekaligus mengurangi book_stock (sama seperti add_invoice)

michaelr1300 commented on application/modules/invoice/controllers/Invoice.php:
```diff
@@ -192,12 +189,39 @@ public function edit($invoice_id)
+                }
+            }
+            // Hapus Transaction yang sudah ada
+            $this->db->where('invoice_id', $invoice_id)->delete('book_transaction');
```
Sama seperti edit invoice_book, tapi ada tambahan step untuk update semua book_transaction (book_id nya sama) yang lebih baru.

Misalkan ada buku dengan stock awal 450, kemudian ada 3 book_transaction sehingga stock akhir jadi 50 (Gambar 1)
![image](https://user-images.githubusercontent.com/46708593/118375479-1ab65b80-b5ec-11eb-9593-ed50084e5b65.png)


Lalu terjadi perubahan lewat edit_invoice untuk transaksi yang tengah, sehingga transaksi tengah jadi -50 (sebelumnya -100).

Jika book_transaction yang lebih baru (transaksi atas) tidak diupdate saat ada perubahan (hanya hapus yang tengah lalu tambah yang baru), maka data stock_awal stock_akhir menjadi tidak konsisten (Gambar 2)
![image](https://user-images.githubusercontent.com/46708593/118375495-2f92ef00-b5ec-11eb-8371-ac21d877bd1f.png)


Code ini berfungsi untuk menjaga data stock_awal dan stock_akhir agar tetap konsisten meski dilakukan perubahan. Skenario yang sama tapi dengan code ini (Gambar 3 & 4)
![image](https://user-images.githubusercontent.com/46708593/118375813-5eaa6000-b5ee-11eb-892d-cf772aa5eb9c.png)

Stock awal dan stock akhir terupdate jadi tetap konsisten
![image](https://user-images.githubusercontent.com/46708593/118375818-69fd8b80-b5ee-11eb-85c8-da1663bea6c1.png)

bagaskarala reviewed (approved):
nice aku coba sudah sesuai mutasi transaksinya
